### PR TITLE
Expand elliptic curves supported for TLS ephemeral ECDH key (#23083)

### PR DESCRIPTION
**Pull request by rshearman**
When the code was originally OpenSSL didn't have any support for
automatically using one or a list of elliptic curves for the TLS
ephemeral ECDH key, and instead required specifying one manually and
P-256 was a reasonable choice.

However, now next-generation cryptography guidance, such as CNSA
Suite, recommends using ECC P-384 instead of P-256, so to provide
interoperability with peers that only support ECC P-384, or more
secure, then make OpenSSL use automatic curve selection out of a
pre-defined list of appropriate curves. For OpenSSL 1.0.2 this is done
by calling SSL_CTX_set_ecdh_auto, but for OpenSSL 1.1.0 and newer this
is done just by not calling SSL_CTX_set_tmp_ecdh to force a particular
curve.

The default groups in OpenSSL and BoringSSL are all curves of equal or
greater in bits of security than P-256, so using the extended set of
groups doesn't open up an opportunity for a less secure curve to be
used.

@donnadionne

### CODE REVIEW

File: src/core/tsi/ssl_transport_security.cc [hunk 1]
```diff
@@ -829,14 +829,26 @@ static tsi_result populate_ssl_context(
...
       gpr_log(GPR_ERROR, "Could not set ephemeral ECDH key.");
       EC_KEY_free(ecdh);
       return TSI_INTERNAL_ERROR;
     }
-    SSL_CTX_set_options(context, SSL_OP_SINGLE_ECDH_USE);
     EC_KEY_free(ecdh);
+#elif OPENSSL_VERSION_NUMBER < 0x10100000
+    /* OpenSSL 1.0.2 and newer can automatically support ECDH and use
+       the most appropriate parameters */
+    SSL_CTX_set_ecdh_auto(ctx->ctx, 1);
```
Comment (yihuazhang) [inline src/core/tsi/ssl_transport_security.cc:845]: Should we handle the case when it returns 0?

**Comment by yihuazhang** [comment]
Could you please address EasyCLA issue as well?

File: src/core/tsi/ssl_transport_security.cc (hunk 1 above)
Comment (rshearman) [reply src/core/tsi/ssl_transport_security.cc:845]: Yes, good point. Will fix.

**Comment by rshearman** [comment]
> Could you please address EasyCLA issue as well?

Noted and in progress.

File: src/core/tsi/ssl_transport_security.cc (hunk 1 above)
Comment (rshearman) [reply src/core/tsi/ssl_transport_security.cc]: The latest revision adds this check and handles the case where SSL_CTX_set_ecdh_auto fails.

**Review by yihuazhang — COMMENTED** [review]
LGTM. Will approve it once confirming tests are green.

**Pull request closed without merging**